rust/src/modbus/detect.rs

```diff
@@ -0,0 +1,1345 @@
+                    '>' => Ok(num..std::u16::MAX),
+                    '<' => Ok(std::u16::MIN..num),
+                    _ => Ok(num..num),
+                },
+                Err(_) => {
+                    SCLogError!("Invalid min number: {}", min_str);
+                    Err(())
+                }
+            }
+        } else {
+            Err(())
+        }
+    } else {
+        let min = match min_str.parse::<u16>() {
+            Ok(num) => num,
+            Err(_) => {
+                SCLogError!("Invalid min number: {}", min_str);
+                return Err(());
+            }
+        };
+
+        let max = match max_str.parse::<u16>() {
+            Ok(num) => num,
+            Err(_) => {
+                SCLogError!("Invalid max number: {}", max_str);
+                return Err(());
+            }
+        };
+
+        Ok(min..max)
+    }
+}
+
+/// Intermediary function between the C code and the access type parsing function.
+#[no_mangle]
+pub unsafe extern "C" fn rs_parse_access(c_arg: *const c_char) -> *mut c_void {
+    if c_arg.is_null() {
+        return std::ptr::null_mut();
+    }
+    if let Ok(arg) = CStr::from_ptr(c_arg).to_str() {
+        match parse_access(&arg) {
+            Ok(detect) => return Box::into_raw(Box::new(detect)) as *mut c_void,
+            Err(()) => return std::ptr::null_mut(),
+        }
+    }
+    std::ptr::null_mut()
+}
+
+/// Intermediary function between the C code and the function code parsing function.
+#[no_mangle]
+pub unsafe extern "C" fn rs_parse_function(c_arg: *const c_char) -> *mut c_void {
+    if c_arg.is_null() {
+        return std::ptr::null_mut();
+    }
+    if let Ok(arg) = CStr::from_ptr(c_arg).to_str() {
+        match parse_function(&arg) {
+            Ok(detect) => return Box::into_raw(Box::new(detect)) as *mut c_void,
+            Err(()) => return std::ptr::null_mut(),
+        }
+    }
+    std::ptr::null_mut()
+}
+
+/// Intermediary function between the C code and the unit ID parsing function.
+#[no_mangle]
+pub unsafe extern "C" fn rs_parse_unit_id(c_arg: *const c_char) -> *mut c_void {
+    if c_arg.is_null() {
+        return std::ptr::null_mut();
+    }
+    if let Ok(arg) = CStr::from_ptr(c_arg).to_str() {
+        match parse_unit_id(&arg) {
+            Ok(detect) => return Box::into_raw(Box::new(detect)) as *mut c_void,
+            Err(()) => return std::ptr::null_mut(),
+        }
+    }
+    std::ptr::null_mut()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_modbus_free(ptr: *mut c_void) {
+    if !ptr.is_null() {
+        let _ = Box::from_raw(ptr as *mut DetectModbusRust);
+    }
+}
+
+/// Compares a transaction to a signature to determine whether the transaction
+/// matches the signature. If it does, 1 is returned; otherwise 0 is returned.
+#[no_mangle]
+pub extern "C" fn rs_modbus_inspect(tx: &ModbusTransaction, modbus: &DetectModbusRust) -> u8 {
+    let req = match &tx.request {
+        Some(r) => r,
+        None => return 0,
```

Comment:
So we can only run inspection if we saw a request? Since the parser has indicated it accepts GAPS, we might see response only transactions, right? Both in packet loss and midstream cases. So I wonder if there is anything we could salvage from just the responses that is worth inspecting here?

Reply:
We can certainly improve inspection.

I assume you mean that we passed the flag `APP_LAYER_PARSER_OPT_ACCEPT_GAPS`. I'm not sure what exactly this flag does because I can't find documentation for it.

The C parser did not accept gaps. The rust parser can accept gaps but messages must be encapsulated in a single packet. E.g. given,
- packet 1: message 1
- packet 2: first half of message 2
- packet 3: second half of message 2
- packet 4: message 3
- packet 5: message 4

This will occur:
- Losing no packets -- parsing packet 2 will result in an incomplete error (`AppLayerResult::incomplete(_)`).
- Losing only packet 2 or only packet 3 -- will result in an unrecoverable error (`AppLayerResult::err()`).
- Losing packet 4 -- would be fine.

We are considering adding more complex gap handling in the future, allowing loss of packet 3 to output a partial packet 2 and recover in time for packet 4 and 5.